pkg/build/docker_generic.go

```diff
@@ -26,43 +26,60 @@ type DockerGenericBuilder struct {
 
 type DockerGenericBuilderConfig struct {
 	// Custom base path where we find the test source
-	Path      string             `toml:"path" default:"./"`
-	BuildArgs map[string]*string `toml:"build_args"` // ok if nil
+	Path       string             `toml:"path"`
+	Dockerfile string             `toml:"dockerfile"`
+	BuildArgs  map[string]*string `toml:"build_args"` // ok if nil
 }
 
-// Build builds a testplan written in Go and outputs a Docker container.
-func (b *DockerGenericBuilder) Build(ctx context.Context, in *api.BuildInput, ow *rpc.OutputWriter) (*api.BuildOutput, error) {
-	cfg, ok := in.BuildConfig.(*DockerGenericBuilderConfig)
+func newDockerGenericBuilderConfig(buildConfig interface{}) (*DockerGenericBuilderConfig, error) {
+	cfg, ok := buildConfig.(*DockerGenericBuilderConfig)
+
 	if !ok {
-		return nil, fmt.Errorf("expected configuration type DockerGenericBuilderConfig, was: %T", in.BuildConfig)
+		return nil, fmt.Errorf("expected configuration type DockerGenericBuilderConfig, was: %T", buildConfig)
 	}
 
-	cliopts := []client.Opt{client.FromEnv, client.WithAPIVersionNegotiation()}
-
-	var (
-		basesrc  = in.UnpackedSources.BaseDir
-		cli, err = client.NewClientWithOpts(cliopts...)
-	)
-	if err != nil {
-		return nil, err
+	if len(cfg.Path) == 0 {
+		cfg.Path = "./"
 	}
 
-	planPath := cfg.Path
-	basePathForPlan := path.Join("/plan", planPath)
+	if len(cfg.Dockerfile) == 0 {
+		cfg.Dockerfile = "Dockerfile"
+	}
 
 	if cfg.BuildArgs == nil {
 		cfg.BuildArgs = make(map[string]*string)
 	}
 
-	if _, ok = cfg.BuildArgs["PLAN_PATH"]; !ok {
+	if _, ok := cfg.BuildArgs["PLAN_PATH"]; !ok {
 		cfg.BuildArgs["PLAN_PATH"] = &cfg.Path
 	}
 
+	return cfg, nil
+}
+
+// Build builds a testplan written in Go and outputs a Docker container.
+func (b *DockerGenericBuilder) Build(ctx context.Context, in *api.BuildInput, ow *rpc.OutputWriter) (*api.BuildOutput, error) {
+	cfg, err := newDockerGenericBuilderConfig(in.BuildConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cliopts := []client.Opt{client.FromEnv, client.WithAPIVersionNegotiation()}
+
+	basesrc := in.UnpackedSources.BaseDir
+	cli, err := client.NewClientWithOpts(cliopts...)
+	if err != nil {
+		return nil, err
+	}
+
+	planPath := cfg.Path
+	basePathForPlan := path.Join("/plan", planPath)
+
 	opts := types.ImageBuildOptions{
 		Tags:        []string{in.BuildID},
 		BuildArgs:   cfg.BuildArgs,
 		NetworkMode: "host",
-		Dockerfile:  filepath.Join(basePathForPlan, "Dockerfile"),
+		Dockerfile:  filepath.Join(basePathForPlan, cfg.Dockerfile),
```

Comment:
It looks like this was the core change; why do you need to introduce `newDockerGenericBuilderConfig`?